batchflow/plot.py

```diff
@@ -0,0 +1,830 @@
+
+import numpy as np
+
+from IPython.display import display
+from matplotlib import patheffects
+from matplotlib import pyplot as plt
+from matplotlib.colors import ColorConverter, ListedColormap, is_color_like
+from matplotlib.lines import Line2D
+from matplotlib.patches import Patch
+from mpl_toolkits import axes_grid1
+
+from .utils import to_list
+
+
+
+class LoopedList(list):
+    """ List that loops from given position (default is 0).
+
+        Examples
+        --------
+        >>> l = LoopedList(['a', 'b', 'c'])
+        >>> [l[i] for i in range(9)]
+        ['a', 'b', 'c', 'a', 'b', 'c', 'a', 'b', 'c']
+
+        >>> l = LoopedList(['a', 'b', 'c', 'd'], loop_from=2)
+        >>> [l[i] for i in range(9)]
+        ['a', 'b', 'c', 'd', 'c', 'd', 'c', 'd', 'c']
+
+        >>> l = LoopedList(['a', 'b', 'c', 'd', 'e'], loop_from=-1)
+        >>> [l[i] for i in range(9)]
+        ['a', 'b', 'c', 'd', 'e', 'e', 'e', 'e', 'e']
+    """
+    def __init__(self, *args, loop_from=0, **kwargs):
+        self.loop_from = loop_from
+        super().__init__(*args, **kwargs)
+
+    def __getitem__(self, idx):
+        if idx >= len(self):
+            pos = self.loop_from + len(self) * (self.loop_from < 0)
+            if pos < 0:
+                raise IndexError(f"List of length {len(self)} is looped from {self.loop_from} index")
+            idx = pos + (idx - pos) % (len(self) - pos)
+        return super().__getitem__(idx)
+
+# pylint: disable=invalid-name
+class preprocess_and_imshow:
+    """ TODO """
+    def __init__(self, ax, array, *args, mask_values=(), order_axes=None, vmin=None, vmax=None, **kwargs):
+        self.mask_values = to_list(mask_values) if mask_values is not None else []
+        self.order_axes = order_axes
+        self.vmin, self.vmax = vmin, vmax
+
+        new_array = self._preprocess(array)
+        self.im = ax.imshow(new_array, *args, vmin=vmin, vmax=vmax, **kwargs)
+
+    def _preprocess(self, array):
+        masks = [array == m if isinstance(m, Number) else m(array) for m in self.mask_values]
+        mask = reduce(np.logical_or, masks, np.isnan(array))
+        new_array = np.ma.array(array, mask=mask)
+
+        order_axes = self.order_axes[:array.ndim]
+        new_array = np.transpose(new_array, order_axes)
+        return new_array
+
+    def set_data(self, array):
+        """ TODO """
+        vmin_new = np.nanmin(array) if self.vmin is None else self.vmin
+        vmax_new = np.nanmax(array) if self.vmax is None else self.vmax
+        clim = [vmin_new, vmax_new]
+        self.im.set_clim(clim)
+
+        new_array = self._preprocess(array)
+        self.im.set_data(new_array)
+
+    def __getattr__(self, key):
+        if self.im is None:
+            return getattr(self, key)
+        return getattr(self.im, key)
+
+    def __repr__(self):
+        if self.im is None:
+            return super().__repr__()
+        return self.im.__repr__()
+
+
+class plot:
+    """ Plotting backend for matplotlib.
```

Comment:
Mention the ways to display data (`imshow`, `wiggle`...). The name `plot` might suggest that it's all about line plotting.